plugins/org.python.pydev.parser/src/org/python/pydev/parser/grammar_fstrings/FStringsGrammarFactory.java

```diff
@@ -0,0 +1,25 @@
+package org.python.pydev.parser.grammar_fstrings;
+
+import org.python.pydev.parser.jython.FastCharStream;
+
+public class FStringsGrammarFactory {
+
+    public static FStringsGrammar createGrammar(String s) {
+
+        while (s.contains("\\N{")) {
```

Comment:
You could do something as:

```
int i;
while((i=s.indexOf("\\N{")) != -1) {
```

to avoid having to iterate more than once (one for `contains` and another one for the `indexOf`). Still, there are some issues in the code... the code below doesn't check if `s_.indexOf("}");` returns `-1` (so, if you have`\\N{` without closing it you could have issues) and the `s = s.replace(s.charAt(pos), '_');` that's done will actually replace all occurrences of `{`, which is not what you want (so, if your input has multiple `\\N` with code that should be  replaced as `{var}`, it won't find the `var` (please create test cases for these situations)).

Also, I don't think you need to create that many temporary strings, a single loop could probably do what you want... Probably something as the code below would be a bit better (note: I haven't run it, so, please update the code accordingly if something is wrong / tests fail -- it's mostly to provide the idea for the code).

```
        char[] c = s.toCharArray();
        for (int i = 0; i < c.length; i++) {
            if (c[i] == '\\' && i + 2 < c.length && c[i + 1] == 'N' && c[i + 2] == '{') {
                c[i + 2] = '_';
                i += 2;
                for (; i < c.length; i++) {
                    if (c[i] == '}') {
                        c[i] = '_';
                        break;
                    }
                }
            }
        }
```

You should also test in conditions which could have corner cases, such as just `\N` or `\N{` or `\N{\N{}`.